### PR TITLE
Topic recognition - s4753820 ViT ADNI Brain Data Classification 

### PR DESCRIPTION
**Pull request by Antimatter543**
Hi,

This pull request integrates a solution for Alzheimer's Disease detection using Vision Transformers (ViT), applying it to the ADNI dataset. The goal was to classify MRI brain scans into two categories: Alzheimer's Disease (AD) and Healthy (No Alzheimer's). The implementation leverages PyTorch and a Vision Transformer architecture, combined with aggressive data augmentation and a comprehensive training, validation, and testing pipeline. While we aimed for a target test accuracy of 80%, our final model reached a test accuracy of 67.93% which unfortunately means our model architecture/hyperparameters likely weren't tuned enough.

### Contents / Files Added:
- dataset.py: Contains the functions for dataset loading, augmentation, and splitting the dataset into training, validation, and test sets.

- modules.py: Defines the Vision Transformer model architecture, including patch embedding, transformer layers, and classification heads.

- train.py: Implements the full training loop, including validation steps, learning rate scheduling, checkpoint saving, and loss/accuracy plotting.

- predict.py: Demonstrates the usage of the trained model for prediction. It prints classification results and provides visualizations such as confusion matrices and image grids with predicted labels.

### Documentation:
README.md: A more detailed explanation of the project, including problem description, algorithm explanation, architecture overview, preprocessing steps, example usage, and experiment results.

Let me know if any changes are required or feedback needs to be incorporated!
Sincerely,
Yuvraj Fowdar
s4753820

### CODE REVIEW

**Comment by prprakash02** [comment]
**This is an initial inspection**

**1) Recognition Problem:**

a) Algorithm solves the respective pattern recognition problem appropriately: **(+3)**
b) Algorithm implemented in PyTorch demonstrated and functions as intended to solve the respective problem: **(+3)**
c) Good design implemented in PyTorch that solves the problem: **(+1)**
d) Commenting: **(+1)**
e) Algorithm is normal or above difficulty while implemented in PyTorch while solving the respective problem: Hard level **(+5)**
f) Algorithm is hard difficulty while implemented in PyTorch while solving the respective problem: Hard level **(+5)**

**(Total: 18/20)**

**2) Commit Log:**

a) Meaningful commit messages with evidence of individual work and originality: **(+2)**
b) Progressive commits and logical log structure providing evidence of individual work and originality: **(+3)**

**(Total: 5/5)**

**3) Documentation:**

Requirement:

Tittle: Done
Description of the algorithm: Done
Problem that it solves: Done
How it works in a paragraph (example any pre-processing): Done
Figure/visualisation (any input, output & results): Done
Dependencies: Done

Marking:

a) Description and explanation of the working principles of the algorithm implemented and the problem it solves: **(+5)**
b) Description of usage and comments throughout scripts: **(+3)**
c) Proper formatting using GitHub markdown: **(+2)**

**(Total: 10/10)**

**4) Pull Request**

a) Creating a pull-request into correct branch with a working and demonstrable version of the algorithm: **(+2)**
b) Incorporating feedback into the pull request: **(+2)**
c) Description of and comments within the pull request: **(+1)**

**(Total: 5/5)**

**Feedback:**

- Good attempt. 
- I think one of the major issue lies in data pre-processing, where test and validation sets were used **after** data augmentation. 
- In CV tasks, the training dataset can be modified (augmented) for various reasons but test/valid should be left untouched

**Any Changes Needed: No**

**Comment by aniketgupta17** [comment]
Observational Feedback

Pull Request:
Correctly created the Pull request from Topic Recognition Branch .
The pull request includes a clear description about the file structure .


File Organizing: Well-organized files.

Commit Log:
Commit messages are progressive for the Recognition Problem solved using 4 files .

Commits are regularly made.

Documentation:
Code comments and docstrings are included.
Proper GitHub markdown formatting is used, with organized headings, lists, and code blocks.
Should  add about Conclusion and future improvements in the ReadMe.

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>Header blocks. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>OK solution to problem. </td><td>-2</td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Hard. ViT (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Good. </td><td></td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). On time submission.</td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-2</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_
